docs/Color-Picker/Color-Picker.md

```diff
@@ -1 +1,285 @@
+  ...
+}
+```
+
+### After Event
+
+#### Connect
+Send `AUTHENTICATE`
+```js
+{
+  token: String,
+}
+```
+
+#### Start a new listen
+Send `REQUEST_CREATE_LISTEN`
+```js
+{
+  songId: String,
+  timestamp: String,
+}
+```
+
+**Response**
+```js
+{
+  id: String,
+  songId: String,
+  timestamp: String,
+}
+````
+
+#### Choose a color
+Send `COLOR_CHOSEN`
+```js
+{
+  listenId: String,
+  color: String,
+  timeOffset: Number,
+}
+```
+
+#### Get list of songs
+Send `REQUEST_GET_SONGS`
+
+**Response**
+```js
+[
+  {
+    id: String,
+    mediaUrl: String,
+    title: String,
+    artist: String,
+    length: Number,
+  }
+]
+```
+
+#### Get list of listens for a song
+Send `REQUEST_GET_LISTENS`
+```js
+{
+  songId: String,
+}
+```
+
+**Response**
+```js
+[
+  {
+    id: String,
+    songId: String,
+    timestamp: String,
+    choices: [
+      {
+        timeOffset: Number,
+        color: String,
+      }
+    ],
+  },
+  ...
+]
+```
+
+#### Get list of choices for a listen
+Send `REQUEST_GET_CHOICES`
+```js
+{
+  listendId: String,
+}
+```
+
+**Response**
+```js
+[
+  {
+    listenId: String,
+    color: String,
+    timeOffset: Number,
+  },
+  ...
+]
+```
+
+## Data Storage
+
+### During Event
+
+#### Song information
+```js
+{
+  songId: String,           // Partition key
+  listenId: 'SONG_INFO',    // Range key
+  title: String,            // Helpful for in-app displays
+  startTime: String,
+  endTime: String,
+}
+```
+
+#### User choices
+```js
+{
+  songId: String,           // Can be set to 'CONCERT_1' for the entire event
+  listenId: String,         // equal to userId
+  choices: {
+    [timestamp]: color,     // Key is the timestamp of the choice
+    Number: String,         // Value is the color choice
+  },
+}
+```
+
+#### Aggregated choices
+```js
+{
+  songId: String,
+  listenId: 'AGGREGATE',
+  choices: {
+    [timestamp]: {
+      COLOR_RED: Number,
+      COLOR_BLUE: Number,
+      COLOR_GREEN: Number,
+      ...
+    },
+    Number: Map,
+  },
+}
+```
+
+### After Event
+
+#### Song information
+```js
+{
+  songId: String,           // Partition key
+  listenId: 'SONG_INFO',    // Range key
+  mediaUrl: String,         // Cached by app for playback
+  title: String,            // Helpful for in-app displays
+  artist: String,           // Helpful for in-app displays
+  length: Number,           // Helpful for in-app displays
+}
+```
+
+#### User choices
+```js
+{
+  songId: String,
+  listenId: String,         // Compound key: userId$timestamp
+  choiceType: String,       // CHOICE_COLOR | CHOICE_EMOTION
+  choices: {
+    [timeOffset]: choice,   // Key is the number of milliseconds since the starting time of this song during the event
+    Number: String,         // Value is the color or emotion choice
+  },
+}
+```
+
+#### Aggregated choices
+```js
+{
+  songId: String,
+  listenId: 'AGGREGATE',
+  choices: {
+    [timeOffset]: {
+      COLOR_RED: Number,
+      COLOR_BLUE: Number,
+      COLOR_GREEN: Number,
+      ...
```

Comment:
I'm guessing that since we may have emotions as well as colors at this point, the choices list here may contain keys for emotions in addition to keys for colors. Is that right?

Reply:
Emotion was the next thing I was going to weave in, actually wanted to ask your opinion. Do you think there should be a `choiceType` field (CHOICE_COLOR, CHOICE_EMOTION) and have any fields I might have called "color" be called "choice"?

Reply:
If color and emotion selection data will be sharing a table, my gut is that it might be best practice to have an attribute (e.g. choiceType like you've suggested above) that explicitly says which type of data any given entry is. Honestly, though, unless there's a big performance difference between querying based on attribute starts with vs. querying on attribute equals (choiceType == CHOICE_COLOR vs key starts with COLOR_), I don't know that it matters a ton. Fine leaving it whichever way you prefer.

Reply:
(Separate, but maybe related: if you're working on emotion API contracts as well as data storage, I'd suggest that even though emotion and color selections will share a bunch of the same fields, it probably makes sense to stand up separate API's rather than extend the existing ones. Improves readability, makes it easier to keep track of necessary test coverage, etc)

Reply:
> If color and emotion selection data will be sharing a table, my gut is that it might be best practice to have an attribute (e.g. choiceType like you've suggested above) that explicitly says which type of data any given entry is. 

I agree, let's do that.

As for separate API's, I could be overlooking something, but I don't think it will be necessary. Seems like just a couple if statements to me, but I think we can call that one as we develop it.

Reply:
Sure. No big deal either way, especially for a project this small